### ISSUE TITLE
Support SQLite columns having no data type declaration

### ISSUE THREAD

**Issue by mrigger** [jOOQ/jOOQ#12788]
Not sure if the following is an issue and/or already tracked; https://github.com/jOOQ/jOOQ/issues/179 is a related issue about type affinities in SQLite.

### Expected behavior

I would expect that SQLite tables that have columns with no data type declaration can be processed by jOOQ such as the example shown below:

```sql
SQLite version 3.38.0 2021-12-16 19:50:02
Enter ".help" for usage hints.
sqlite> CREATE TABLE t0(c0);
```

Columns with unknown data types work as expected. For example, creating a table `CREATE TABLE t0(c0 asdf);`, with the non-existing data type `asdf` is processed without errors and results in generic signatures being generated. Perhaps this could be a reasonable default behavior?

### Actual behavior

I executed `org.jooq.codegen.GenerationTool` with the configuration below:
```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<configuration xmlns="http://www.jooq.org/xsd/jooq-codegen-3.16.0.xsd">
  <!-- Configure the database connection here -->
  <jdbc>
    <driver>org.sqlite.JDBC</driver>
    <url>jdbc:sqlite:bugs.db</url>
    <user></user>
    <password></password>
  </jdbc>

  <generator>
    <name>org.jooq.codegen.JavaGenerator</name>

    <database>
      <name>org.jooq.meta.sqlite.SQLiteDatabase</name>
      <includes>.*</includes>
      <excludes></excludes>
    </database>

    <target>
      <!-- The destination package of your generated classes (within the destination directory) -->
      <packageName>org.sqlancer.jooqtest</packageName>

      <!-- The destination directory of your generated classes. Using Maven directory layout here -->
      <directory>gen</directory>
    </target>
  </generator>
</configuration>
```

Unexpectedly, jOOQ fails with the following error:
```sql
16:33:14 INFO Cannot parse SQL: CREATE TABLE t0(c0
)
org.jooq.impl.ParserException: Keyword 'AS' expected: [2:2] CREATE TABLE t0(c0
)[*]
	at org.jooq.impl.DefaultParseContext.expected(ParserImpl.java:13928)
	at org.jooq.impl.DefaultParseContext.parseKeyword(ParserImpl.java:13287)
	at org.jooq.impl.DefaultParseContext.parseCreateTable(ParserImpl.java:4349)
	at org.jooq.impl.DefaultParseContext.parseCreate(ParserImpl.java:2758)
	at org.jooq.impl.DefaultParseContext.parseQuery(ParserImpl.java:1053)
	at org.jooq.impl.DefaultParseContext.lambda$parseQuery0$2(ParserImpl.java:892)
	at org.jooq.impl.DefaultParseContext.wrap(ParserImpl.java:14140)
	at org.jooq.impl.DefaultParseContext.parseQuery0(ParserImpl.java:892)
	at org.jooq.impl.ParserImpl.parseQuery(ParserImpl.java:713)
	at org.jooq.impl.ParserImpl.parseQuery(ParserImpl.java:708)
	at org.jooq.meta.sqlite.SQLiteTableDefinition.interpretedTable(SQLiteTableDefinition.java:95)
	at org.jooq.meta.sqlite.SQLiteTableDefinition.getElements0(SQLiteTableDefinition.java:147)
	at org.jooq.meta.AbstractElementContainerDefinition.getElements(AbstractElementContainerDefinition.java:93)
	at org.jooq.meta.AbstractTableDefinition.getColumns(AbstractTableDefinition.java:186)
	at org.jooq.codegen.JavaGenerator.generateTable(JavaGenerator.java:5350)
	at org.jooq.codegen.JavaGenerator.generateTables(JavaGenerator.java:5337)
	at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:621)
	at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:576)
	at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:491)
	at org.jooq.codegen.GenerationTool.run0(GenerationTool.java:898)
	at org.jooq.codegen.GenerationTool.run(GenerationTool.java:239)
	at org.jooq.codegen.GenerationTool.generate(GenerationTool.java:234)
	at org.jooq.codegen.GenerationTool.main(GenerationTool.java:206)
```

Having columns without data types is not uncommon in SQLite. For example, creating an `AUTOINCREMENT` column (e.g., `CREATE TABLE t0(c0 INTEGER PRIMARY KEY AUTOINCREMENT);`) implicitly creates a table (`CREATE TABLE sqlite_sequence(name,seq)`) that fails with the same error.

### Versions

I used jOOQ 3.16.0.

**Comment by lukaseder** [comment]
Thanks for your report, @mrigger. Are you going to run your sqlancer tests against the jOOQ parser? That will be a fun exercise, surely you'll discover quite a few bugs! :)

We're parsing SQLite's table definitions to discover meta data, like constraints, which cannot really be discovered otherwise using `pragma` commands. Obviously, this means we have to support whatever SQLite thinks is valid SQL. In this case, I guess we can just assume `SQLDataType.OTHER`

**Comment by lukaseder** [comment]
`ALTER TABLE t ADD c` also works ("obviously")

**Comment by lukaseder** [comment]
So far, our parser assumes that if column types are ommitted, we're parsing a `CREATE TABLE AS SELECT` statement (which doesn't support such column lists in SQLite, but we do)...

**Comment by lukaseder** [comment]
SQLite 😅

What's the difference between these?

```sql
create table t (a primary key);
create table t (a primaary key);
```

**Comment by mrigger** [comment]
I have been looking for students to work on extending SQLancer to test jOOQ in a Bachelor thesis project, but so far nobody decided to work on this project. I've just started playing with jOOQ to find some inspiration for future research. ;)

**Comment by mrigger** [comment]
> SQLite sweat_smile
> 
> What's the difference between these?
> 
> ```sql
> create table t (a primary key);
> create table t (a primaary key);
> ```

I was wondering why jOOQ does not translate `STRING` data-type columns to Java string types, and finally remembered that the data type is called `TEXT` in SQLite...

**Comment by lukaseder** [comment]
Great to hear! Hopefully, you'll find a student to work on this :)

Anyway. I'm not sure if it makes sense to emulate all of SQLite's madness in jOOQ's parser (see above example). Any token that isn't a recognised keyword is a "data type", SQLite really doesn't care. Even whitespace in "type names" makes the thing still part of the type. So, this short form you're suggesting here will only be accepted if the parser recognises a `,` or `)` token, which have clear semantics in this context, even in SQLite.

If someone actually needs `create table t (a primary key)` to parse, we can futher discuss, in a new issue.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0 and 3.16.1 (#12792). Thanks again for your report!

**Comment by mrigger** [comment]
Wow, that was quick. Thanks!
